File: Cpp/211_Design_Add_and_Search_Words_Data_Structure.cpp
class WordDictionary {
public:
    struct TrieNode {
        bool is_leaf;
        vector<TrieNode*> children;

        TrieNode() : is_leaf(false), children(26, nullptr){
        }
    };
    
    /** Initialize your data structure here. */
    WordDictionary() : root(new TrieNode()){
        
    }
    
    /** Adds a word into the data structure. */
    void addWord(string word) {
        TrieNode* node = root;
        for(auto& ch : word){
            int index = ch -'a';
            if(node->children[index] == nullptr){
                node->children[index] = new TrieNode();
            }
            node = node->children[index];
        }
        node->is_leaf = true;
    }
    
    /** Returns if the word is in the data structure. A word could contain the dot character '.' to represent any one letter. */
    bool search(string word) {
        return recursive(word, 0, root);
    }
    
    bool recursive(string word, int index, TrieNode* node){
        if(!node) return false;
        if(index == word.size()) return node->is_leaf;
        if(word[index] != '.') return recursive(word, index + 1, node->children[word[index]-'a']);
        for(int i=0; i<26; ++i){
            if (recursive(word, index + 1, node->children[i])) return true;
        }
        return false;
    }
    
private:
    TrieNode* root;
};

/**
 * Your WordDictionary object will be instantiated and called as such:
 * WordDictionary* obj = new WordDictionary();
 * obj->addWord(word);
 * bool param_2 = obj->search(word);
 */

File: Kotlin/211_Design_Add_and_Search_Words_Data_Structure.kt
data class TrieNode(
    val children: Array<TrieNode?> = arrayOfNulls(26),
    var isWord: Boolean = false)

class WordDictionary() {
    val root = TrieNode()

    fun addWord(word: String) {
        var current = root

        for(ch in word) {
            if(current.children[ch - 'a'] == null) current.children[ch - 'a'] = TrieNode()
            current = current.children[ch - 'a']!!
        }

        current.isWord = true
    }

    fun search(word: String): Boolean {
        return recursive(word, 0, root)
    }
    
    fun recursive(word: String, index: Int, node: TrieNode?): Boolean {
        if(node == null) return false
        if(index == word.length) return node.isWord
        if(word[index] != '.') return recursive(word, index + 1, node?.children[word[index] - 'a'])
        
        for(i in 0 until 26){
            if(recursive(word, index + 1, node?.children[i])) return true
        }
        
        return false
    }

}

/**
 * Your WordDictionary object will be instantiated and called as such:
 * var obj = WordDictionary()
 * obj.addWord(word)
 * var param_2 = obj.search(word)
 */